google/cloud/storage/tests/curl_request_integration_test.cc

```diff
@@ -306,3 +307,49 @@ TEST(CurlRequestTest, WellKnownQueryParameters_Multiple) {
   EXPECT_EQ(0U, args.count("ifGenerationMatch"));
   EXPECT_EQ(0U, args.count("ifMetagenerationNotMatch"));
 }
+
+class MockLogBackend : public google::cloud::LogBackend {
+ public:
+  void Process(google::cloud::LogRecord const& lr) { ProcessWithOwnership(lr); }
+  MOCK_METHOD1(ProcessWithOwnership, void(google::cloud::LogRecord));
+};
+
+/// @test Verify that CurlRequest logs when requested.
+TEST(CurlRequestTest, Logging) {
+  // Prepare the Log subsystem to receive mock calls:
+  auto mock_logger = std::make_shared<MockLogBackend>();
+  google::cloud::LogSink::Instance().AddBackend(mock_logger);
+
+  using testing::_;
+  using testing::Invoke;
+
+  std::string log_messages;
+  EXPECT_CALL(*mock_logger, ProcessWithOwnership(_))
+      .WillRepeatedly(Invoke([&log_messages](google::cloud::LogRecord lr) {
+        log_messages += lr.message;
+        log_messages += "\n";
+      }));
+
+  {
+    storage::internal::CurlRequest request(HttpBinEndpoint() + "/post?foo=bar");
+    request.AddHeader("Accept: application/json");
+    request.AddHeader("charsets: utf-8");
+    request.AddHeader("x-test-header: foo");
+
+    request.PrepareRequest(std::string{"this is some text"}, true);
+    auto response = request.MakeRequest();
+    EXPECT_EQ(200, response.status_code);
+  }
+
+  google::cloud::LogSink::Instance().ClearBackends();
+
+  // Verify the URL, and headers, and payload are logged.
+  EXPECT_THAT(log_messages, HasSubstr("/post?foo=bar"));
+  EXPECT_THAT(log_messages, HasSubstr("x-test-header: foo"));
+  EXPECT_THAT(log_messages, HasSubstr("this is some text"));
+  EXPECT_THAT(log_messages, HasSubstr("curl(Info)"));
+  EXPECT_THAT(log_messages, HasSubstr("curl(Send Header)"));
```

Comment:
I don't understand why these curl(...) lines are in the log. For example, why the literal string "curl(Send Header)"? That almost looks like debug content, not something I'd expect to see in our cloud logging logs.

Reply:
These log messages *are* local, and more akin to debug logs rather than anything else.  The goal is to be able to troubleshoot problems with the library, i.e., the library sending the wrong http request or misinterpreting the response.

They do not go to "cloud logging" (by which I am guessing you mean [this](https://cloud.google.com/logging/docs/)), I guess they could be configured to go there, but that would be weird.

If I misunderstood what you meant just let me know, we can chat at more length later.

Reply:
I see, I just misinterpreted the logging you were talking about. LGTM.